posthog/async_migrations/migrations/0002_events_sample_by.py

```diff
@@ -45,78 +66,74 @@ class Migration(AsyncMigrationDefinition):
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"""
-            INSERT INTO {TEMPORARY_TABLE_NAME}
-            SELECT * 
-            FROM {EVENTS_TABLE}
-            WHERE timestamp < toDateTime64(toYYYYMM(now()) - 1, 6)
-            AND timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
-            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            resumable=True,
-            timeout_seconds=7 * 24 * 60 * 60,  # one week
-        ),
-        AsyncMigrationOperation(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            rollback=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            side_effect=lambda: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
-            side_effect_rollback=lambda: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
-        ),
-        AsyncMigrationOperation(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"""
-            INSERT INTO {TEMPORARY_TABLE_NAME}
-            SELECT * 
-            FROM {EVENTS_TABLE}
-            WHERE timestamp >= (SELECT max(timestamp) FROM {TEMPORARY_TABLE_NAME})""",
-            rollback=f"TRUNCATE TABLE IF EXISTS {TEMPORARY_TABLE_NAME} ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            resumable=True,
-            timeout_seconds=3 * 24 * 60 * 60,  # three days
-        ),
-        AsyncMigrationOperation(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"""
-                RENAME TABLE
-                    {EVENTS_TABLE_NAME} to {BACKUP_TABLE_NAME},
-                    {TEMPORARY_TABLE_NAME} to {EVENTS_TABLE_NAME}
-                ON CLUSTER {CLICKHOUSE_CLUSTER}
-            """,
-            rollback=f"""
-                RENAME TABLE
-                    {EVENTS_TABLE_NAME} to {EVENTS_TABLE_NAME}_failed,
-                    {BACKUP_TABLE_NAME} to {EVENTS_TABLE_NAME}
-                ON CLUSTER {CLICKHOUSE_CLUSTER}
-            """,
-        ),
-        AsyncMigrationOperation(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"OPTIMIZE TABLE {EVENTS_TABLE_NAME} FINAL",
-            rollback="",
-            resumable=True,
-        ),
-        AsyncMigrationOperation(
-            database=AnalyticsDBMS.CLICKHOUSE,
-            sql=f"ATTACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            rollback=f"DETACH TABLE {EVENTS_TABLE_NAME}_mv ON CLUSTER {CLICKHOUSE_CLUSTER}",
-            side_effect=lambda: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", True),
-            side_effect_rollback=lambda: setattr(config, "COMPUTE_MATERIALIZED_COLUMNS_ENABLED", False),
-        ),
-    ]
+    @cached_property
+    def operations(self):
+
+        last_partition_op = [generate_insert_into_op(self._partitions[-1])] if len(self._partitions) > 0 else []
```

Comment:
Was about to this but realized it would be problematic.

Say we have 0 events when this migration is generated, then before it is run, we get some events.

To handle for that I've made it so that we always generate the last op.

If we have 0 events, we'll still get an op to handle events with timestamp > 0.

That ensures we're future proof.